### PR TITLE
WIP: Unicode Support

### PR DESCRIPTION
**Pull request by frabbit**
This PR adds support for 4 abstracts, Utf8, Utf16, Ucs2 and Utf32 to the std library. This PR is still work in progress.

### CODE REVIEW

File: tests/unit/src/unit/UnitBuilder.hx [hunk 1]
```diff
@@ -86,13 +88,55 @@ class UnitBuilder {
...
+					// calling eqAbstract for Int leads to a segmentation fault with cppia, but only when compiled
+					// with -D nocppiaast
+					var t1 = Context.follow(Context.typeof(e1));
+					var t1 = haxe.macro.TypeTools.toString(t1);
+					if (t1 == "Int") {
+						macro eq($e1, $e2);
+					} else {
+						//macro eq($e1, $e2);
+						macro eqAbstract($e1 == $e2, $e1, $e2);
+					}
```
Comment (frabbit) [inline tests/unit/src/unit/UnitBuilder.hx]: @hughsando can you take a look at this, i have to call `eq` when the type is Int for cppia, but only with `-D nocppiaast`. if i don't do that it leads to a segfault, but it's only in this special circumstance.

Comment (hughsando) [reply tests/unit/src/unit/UnitBuilder.hx]: The nocppiaast is for "legacy test mode" and I don't really want to fix bugs in this code.
Basically, you should not need to use nocppiaast on new code.
Is is acceptable to use this define to hack out the test, or use a "legacy eq" method?

Comment (frabbit) [reply tests/unit/src/unit/UnitBuilder.hx]: I can leave it as it is now, but it's totally strange why this happens at all. I mean it just compares integers, i also traced the values, nothing special but at some point, after the first ~90 comparisons it just segfaults.

**Comment by hughsando** [comment]
You can dump out the code with -debug and -D annotate_source and might get
an idea of what the types are doing.  It sounds like it is passing an int
as an object or the other way around and getting confused.  It is also
possible it is to do with a 1-byte bool vs a 4-byte bool.

**Comment by frabbit** [comment]
@hughsando sounds like fun ;)

**Comment by markknol** [comment]
I noticed in this PR that the copyright notice has the wrong years.

**Comment by Aurel300** [comment]
What is the status on this? I see some of the classes in the API, but are they up-to-date?

**Comment by frabbit** [comment]
@markknol will fix this after new year's eve ;)

@Aurel300 the status is still open, i would love to get more feedback before merging.

**Comment by fullofcaffeine** [comment]
_Ping_, any news here?

**Comment by nadako** [comment]
@ncannasse please review

**Comment by ncannasse** [comment]
Yes I had that in a to-review list for too long, sorry about it. Will try to do it soon.

**Comment by Simn** [comment]
Ok, so... this one frustrates me a bit because there's so much effort that went into the void. However, at this point I think there's no value in keeping this PR open. The branch is still there in case somebody wants to pick this up and update it to latest Haxe. I could see this becoming a haxelib at first, but this would require a dedicated maintainer.

I apologize for how this went down @frabbit. Communication was quite bad from our side, but all we can do now is do better in the future.

**Pull request closed without merging**